action/flow/instance/auditevents.go

```diff
@@ -0,0 +1,245 @@
+
+// Returns error for failed flow instance
+func (fe *FlowEventContext) Error() error {
+	return fe.flowInstance.returnError
+}
+
+// TaskEventContext provides access to task instance execution details
+type TaskEventContext struct {
+	taskInstance *TaskInst
+}
+
+// Returns flow name
+func (te *TaskEventContext) FlowName() string {
+	return te.taskInstance.flowInst.Name()
+}
+
+// Returns flow ID
+func (te *TaskEventContext) FlowID() string {
+	return te.taskInstance.flowInst.ID()
+}
+
+// Returns task name
+func (te *TaskEventContext) Name() string {
+	return te.taskInstance.task.Name()
+}
+
+// Returns task type
+func (te *TaskEventContext) Type() string {
+	return te.taskInstance.task.TypeID()
+}
+
+// Returns task status
+func (te *TaskEventContext) Status() Status {
+	return convertTaskStatus(te.taskInstance.status)
+}
+
+// Returns application name
+func (te *TaskEventContext) AppName() string {
+	return app.GetName()
+}
+
+// Returns application version
+func (te *TaskEventContext) AppVersion() string {
+	return app.GetVersion()
+}
+
+// Returns working data of current instance. e.g. key and value of current iteration for iterator task.
+func (te *TaskEventContext) GetWorkingData() map[string]interface{} {
+	attrs := make(map[string]interface{})
+	if te.taskInstance.HasWorkingData() {
+		for name, value := range te.taskInstance.workingData {
+			attrs[name] = value.Value()
+		}
+	}
+	return attrs
+}
+
+// Returns activity input data
+func (te *TaskEventContext) ActivityInput() map[string]interface{} {
+	attrs := make(map[string]interface{})
+	 if te.taskInstance.task.ActivityConfig().GetInputAttrs() != nil && te.taskInstance.inScope != nil {
+		 for name := range te.taskInstance.task.ActivityConfig().GetInputAttrs() {
+		 	inVal, _ := te.taskInstance.inScope.GetAttr(name)
+		 	attrs[name] = inVal
+		 }
+	 }
+	return attrs
+}
+
+// Returns output data for completed activity
+func (te *TaskEventContext) ActivityOutput() map[string]interface{} {
+	attrs := make(map[string]interface{})
+	if te.Status() == Completed && te.taskInstance.task.ActivityConfig().GetOutputAttrs() != nil && te.taskInstance.outScope != nil {
+		for name := range te.taskInstance.task.ActivityConfig().GetOutputAttrs() {
+			outVal, _ := te.taskInstance.outScope.GetAttr(name)
+			attrs[name] = outVal
+		}
+	}
+	return attrs
+}
+
+// Returns error for failed task
+func (te *TaskEventContext) Error() error {
+	return te.taskInstance.returnError
+}
+
+func convertFlowStatus(code model.FlowStatus) Status {
+	switch code {
+	case model.FlowStatusNotStarted:
+		return Created
+	case model.FlowStatusActive:
+		return Started
+	case model.FlowStatusCancelled:
+		return Cancelled
+	case model.FlowStatusCompleted:
+		return Completed
+	case model.FlowStatusFailed:
+		return Failed
+	}
+	return Unknown
+}
+
+func convertTaskStatus(code model.TaskStatus) Status {
+	switch code {
+	case model.TaskStatusNotStarted:
+		return Created
+	case model.TaskStatusEntered:
+		return Scheduled
```

Comment:
TaskStatusEntered, doesn't necessarily imply "Scheduled"

Reply:
@fm-tibco : What would be appropriate state then?

Reply:
It depends... entered is when the first link comes in.  scheduled is when all the incoming links have come in and they are done, so it schedules an eval.  We can fix up statuses if needed.

Reply:
@fm-tibco : Thanks. Lets make it more meaningful. :-)